### PR TITLE
[Inductor][CPP] Turns on inline_inbuilt_nn_modules for CPP GEMM template testing

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132487

**Summary**
The CPP GEMM template testing has been skipped with turning on `inline_inbuilt_nn_modules ` as in https://github.com/pytorch/pytorch/issues/131929.  Since https://github.com/pytorch/pytorch/pull/132334 has landed to fix the issues. Turn on this flag back since it's default.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

**Comment by anijain2305** [comment]
Thanks. I forgot to remove that line in the last PR.

File: test/inductor/test_cpu_select_algorithm.py
Comment (jgong5) [inline test/inductor/test_cpu_select_algorithm.py:55]: Should remove this comment too?

**Review by jgong5 — CHANGES_REQUESTED** [review]
CI failures?

File: test/inductor/test_cpu_select_algorithm.py
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_select_algorithm.py]: Thanks and removed.

**Comment by leslie-fang-intel** [comment]
> CI failures?

Hi @anijain2305 @jgong5,

I would like to explain a little more about the UT failures, since it seems tricky....
Generally, I have saw 2 issues related to below 2 test cases.

* Test Case 1:
```
python -u -m pytest -s -v test/inductor/test_cpu_select_algorithm.py -k test_linear_with_pointwise_dynamic_shapes_batch_size_384_in_features_196_out_features_385_bias_False_epilogue_add_cpu_float32
```

* Test Case 2:
```
python -u -m pytest -s -v test/inductor/test_cpu_select_algorithm.py -k test_linear_with_pointwise_dynamic_shapes_batch_size_384_in_features_196_out_features_384_bias_False_epilogue_add_cpu_float32
```

* Issue 1: test with test case 1

  For the way, we modeling post-op add: https://github.com/pytorch/pytorch/blob/64720f3b89a1316d6e0131b20c43d6402bcce4aa/test/inductor/test_cpu_select_algorithm.py#L111
  
  - With `inline_inbuilt_nn_modules` as False, the `other` in add seems be taken as the module attr and cause the batch size dimension specialized. In this way, we can hit `mkl_linear` fusion and lowering into cpp gemm template with `add` be taken as `epilogue` in backend codegen.
  - With `inline_inbuilt_nn_modules` as ON, the `other` in add is a input parameter and the batch size will be a symbolic symbol. For the mkl-linear, we can't do deal with symbolic batch size, and the new lowering flow will be:
      - Fuse `mm` + `add` into `addmm`. The `other` will be taken as the `inp(bias)`.
      - Lowering `addmm` into gemm template.
  
  I feels the above behavior is as expected, @anijain2305 what do you think? If so, we can change the UT correspondingly.

* Issue 2: discrepancy between Test Case 1 and Test Case 2

The only difference is output features is 385 or 384 (same as batch size).

    * For Test Case 1, we have saw batch size as symbolic

    ```
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]  ===== FROZEN GRAPH =====
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]  /localdisk/leslie/torch_inductor_community/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]     def forward(self, arg1_1: "Sym(s0)", arg2_1: "f32[s0, 196]", arg3_1: "f32[s0, 385]"):
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]          # File: /localdisk/leslie/torch_inductor_community/pytorch/test/inductor/test_cpu_select_algorithm.py:231 in forward, code: return self.epilogue(self.linear(x))
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]         _frozen_param1: "f32[196, 385]" = self._frozen_param1
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]         mm: "f32[s0, 385]" = torch.ops.aten.mm.default(arg2_1, _frozen_param1);  arg2_1 = _frozen_param1 = None
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]         
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]          # File: /localdisk/leslie/torch_inductor_community/pytorch/test/inductor/test_cpu_select_algorithm.py:110 in <lambda>, code: return lambda x: x + other
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]         add: "f32[s0, 385]" = torch.ops.aten.add.Tensor(mm, arg3_1);  mm = arg3_1 = None
    V0802 19:05:22.665000 3118430 torch/_inductor/freezing.py:117] [0/0]         return (add,)
    ```

    * For Test Case 2, we have saw batch size as static

    ```
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]     def forward(self, arg1_1: "f32[384, 196]", arg2_1: "f32[384, 384]"):
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         # No stacktrace found for following nodes
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         _frozen_param2 = self._frozen_param2
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         _frozen_param3 = self._frozen_param3
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]          # File: /localdisk/leslie/torch_inductor_community/pytorch/test/inductor/test_cpu_select_algorithm.py:231 in forward, code: return self.epilogue(self.linear(x))
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         _mkl_linear: "f32[384, 384]" = torch.ops.mkl._mkl_linear(arg1_1, _frozen_param3, _frozen_param2, None, 384);  arg1_1 = _frozen_param3 = _frozen_param2 = None
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]          # File: /localdisk/leslie/torch_inductor_community/pytorch/test/inductor/test_cpu_select_algorithm.py:110 in <lambda>, code: return lambda x: x + other
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         add: "f32[384, 384]" = torch.ops.aten.add.Tensor(_mkl_linear, arg2_1);  _mkl_linear = arg2_1 = None
    V0802 19:04:34.003000 3117875 torch/_inductor/freezing.py:117] [0/0]         return (add,)
    ```

It looks like a bug?

**Comment by anijain2305** [comment]
I think your observation about the + add makes sense to me. When you use +dynamic, torch.compile marks everything dynamic. Taking a step back, if you dont want to mark the batch size dynamic, you could use `torch._dynamo.mark_dynamic` on dimensions to selectively mark dims as dynamic (instead of full blown dynamism).

For issue 2, I don't fully understand the issue. So, let me take a look to understand. Give me a few hours.

**Comment by anijain2305** [comment]
@leslie-fang-intel I am struggling with setting up MKL? Are there any instructions? I have 

* Ran `pip install mkl-static mkl-include`
* Ran `USE_MKL=1 USE_MKLDNN=1 python setup.py develop"

But I get these logs during configure
`Library mkl_intel_lp64: not found`

**Comment by leslie-fang-intel** [comment]
Here is another script to reproduce issue2: https://gist.github.com/leslie-fang-intel/aa58be19808779313cb9aca9730d736e

- Test 1 has symbolic batch size
- Test 2 has static batch size

**Comment by leslie-fang-intel** [comment]
Sync with @anijain2305, the behavior of second issue mentioned above is expected by design, since dynamo wants to reuse symbols by keeping the num of symbols as low as possible.

**Comment by anijain2305** [comment]
@leslie-fang-intel and I debugged this further over slack. Our findings are this

* The discrepancy between N=385 vs N=384 is expected. For N=384,  there are no more symbolic shapes in AOT graph, but  for N=385, batch size is symbolic. This is expected because of a quirk in symbolic shapes infra.

For the test case with N = 384, these are the dimensions M = 384, N = 384, K = 196. The point to note is that M = N. Dynamic shapes infra reuses symbols wherever possible (duck typing) to reduce the number of symbols. And therefore, `s0` is reused and then later get removed.

For the test case with N=385, the dimensions are M = 384, N = 385, and K = 196. 3 different dimensions, 3 different symbols, no removal, resulting in symbolic batch size.

In nutshell, the testcase exposes a quirk (expected though) in dynamic shapes infra. 

I would suggest to rewrite the test cases with M = 16 so that M, N, K are always different, always resulting in symbolic batch size.

**Comment by xmfan** [comment]
I didn't get the chance to relax the heuristic yet, but I did expose it as a config in https://github.com/pytorch/pytorch/pull/129804. You can also decorate tests with `@torch.fx.experimental._config.patch(use_duck_shape=False)`

**Comment by leslie-fang-intel** [comment]
After adding `@torch.fx.experimental._config.patch(use_duck_shape=False)`, the preCI passes, @anijain2305 @jgong5 please help to take a look again.

**Review by anijain2305 — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Comment by PaliC** [comment]
@pytorchbot revert -m "this pr is breaking inductor tests internally" -c "ghfirst"

I think you should be able to repro these in OSS

pytorch/inductor:cpu_select_algorithm - test_linear_with_pointwise_dynamic_shapes_batch_size_384_in_features_196_out_features_385_bias_False_epilogue_add_cpu_float32

stack trace - https://pastebin.com/xwCrxAa8

pytorch/test/inductor:cpu_select_algorithm - test_linear_with_pointwise_dynamic_shapes_batch_size_384_in_features_196_out_features_384_bias_False_epilogue_add_cpu_float32 

stack trace - https://pastebin.com/WjM6AnAj

**Comment by leslie-fang-intel** [comment]
Hi @PaliC,

> I think you should be able to repro these in OSS

The testing shows in https://hud.pytorch.org/pr/132487 are all green. Do I need to use some specific system to reproduce this failure?

**Comment by leslie-fang-intel** [comment]
Hi @PaliC, is the failure you posted running with `is_fbcode == True`?
Since `inline_inbuilt_nn_modules` only turns on when `is_fbcode == False` https://github.com/pytorch/pytorch/blob/313aa151da1b84ccaa7fe0067a683490474f11da/torch/_dynamo/config.py#L374, will it be the root-cause for this failure? I have forced `inline_inbuilt_nn_modules == True` for above testcase, do you think it makes senses @anijain2305?

**Comment by leslie-fang-intel** [comment]
Hi @PaliC @anijain2305, I have changed to force to turn on `inline_inbuilt_nn_modules`. Could you kindly help to import and check the internal CI status?

**Review by anijain2305 — APPROVED** [review]

**Comment by anijain2305** [comment]
ghstack PRs cant be imported internally. I am confident that explicitly turning on the flag should be good enough. Merging for now.

**Comment by jansel** [comment]
Reverted by #133570

**Comment by anijain2305** [comment]
Requesting here to wait for a week before merging this PR. My reland PR (merged in Oss), which this PR depends upon, is going through internal tests and might need to be reverted again.

**Comment by leslie-fang-intel** [comment]
> Requesting here to wait for a week before merging this PR. My reland PR (merged in Oss), which this PR depends upon, is going through internal tests and might need to be reverted again.
> 
> Let's wait for a week for things to be stable.

Thanks for the comment. sure, let's wait for a week to land it.

**Comment by leslie-fang-intel** [comment]
> Requesting here to wait for a week before merging this PR. My reland PR (merged in Oss), which this PR depends upon, is going through internal tests and might need to be reverted again.

> Let's wait for a week for things to be stable.

Hi @anijain2305, since https://github.com/pytorch/pytorch/pull/134136 has landed for more than 2 weeks. I am going to rebase and land this PR. Thanks.

**Comment by anijain2305** [comment]
Yes, please. Thanks for your patience.

**Pull request closed without merging**
